src/main/java/ch/njol/skript/expressions/ExprBookPages.java

```diff
@@ -21,97 +21,165 @@
+public class ExprBookPages extends PropertyExpression<ItemType, String> {
+
 	static {
 		Skript.registerExpression(ExprBookPages.class, String.class, ExpressionType.PROPERTY,
-				"[all] [the] [book] (pages|content) of %itemtypes%",
-				"%itemtypes%'s [book] (pages|content)",
-				"[book] page %number% of %itemtypes%", 
-				"%itemtypes%'s [book] page %number%");
+			"[(all [[of] the]|the)] [book] (pages|content) of %itemtypes%",
+			"%itemtypes%'[s] [book] (pages|content)",
+			"[book] page %number% of %itemtypes%",
+			"%itemtypes%'[s] [book] page %number%");
 	}
-	
-	private static final ItemType bookItem = Aliases.javaItemType("book with text");
-	
-	@SuppressWarnings("null")
-	private Expression<ItemType> book;
+
 	@Nullable
 	private Expression<Number> page;
-	
-	@SuppressWarnings("null")
-	@Nullable
+
 	@Override
-	protected String[] get(Event e) {
-		ItemStack itemStack = book.getSingle(e).getRandom();
-		if (itemStack == null || !bookItem.isOfType(itemStack))
-			return null;
-		List<String> pages = ((BookMeta) itemStack.getItemMeta()).getPages();
-		if (page != null){
-			Number pageNumber = page.getSingle(e);
-			if (pageNumber == null){
-				return null;
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, SkriptParser.ParseResult parseResult) {
+		if (matchedPattern == 0 || matchedPattern == 1) {
+			setExpr((Expression<ItemType>) exprs[0]);
+		} else {
+			if (matchedPattern == 2) {
+				page = (Expression<Number>) exprs[0];
+				setExpr((Expression<ItemType>) exprs[1]);
+			} else {
+				setExpr((Expression<ItemType>) exprs[0]);
+				page = (Expression<Number>) exprs[1];
 			}
-			int page = pageNumber.intValue();
-			if ((page) > pages.size() || page < 1){
-				return null;
+		}
+		return true;
+	}
+
+	@Override
+	@Nullable
+	@SuppressWarnings("null")
+	protected String[] get(Event event, ItemType[] source) {
+		List<String> pages = new ArrayList<>();
+		for (ItemType itemType : source) {
+			ItemMeta meta = itemType.getItemMeta();
+			if (!(meta instanceof BookMeta))
+				continue;
+			BookMeta bookMeta = (BookMeta) meta;
+			if (page == null) {
+				pages.addAll(bookMeta.getPages());
+			} else {
+				Number pageNumber = page.getSingle(event);
+				if (pageNumber == null)
+					continue;
+				int page = pageNumber.intValue();
+				if (page <= 0 || page > bookMeta.getPageCount())
+					continue;
+				pages.add(bookMeta.getPage(page));
 			}
-			return new String[]{pages.get(page - 1)};
-		}else{
-			return pages.toArray(new String[pages.size()]);
 		}
+		return pages.toArray(new String[0]);
 	}
-	
+
 	@Override
-	public boolean isSingle() {
-		return page != null;
+	@Nullable
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		switch (mode) {
+			case SET:
+			case RESET:
+			case DELETE:
+				return CollectionUtils.array(page == null ? String[].class : String.class);
+			case ADD:
+				return page == null ? CollectionUtils.array(String[].class) : null;
```

Comment:
```suggestion
				return CollectionUtils.array(isSingle() ? String.class : String[].class);
			case ADD:
				return isSingle() ? null : CollectionUtils.array(String[].class);
```
I suggest using the already existing isSingle() method to make it easier to understand why this is doing this.